src/util/metadata/global.rs

```diff
@@ -31,171 +30,308 @@ impl MetadataSpec {
+                object,
+                val,
+                mask,
+                ordering,
+            ),
         }
     }
-}
 
-/// A function to atomically compare-and-exchange the specified metadata's content.
-///
-/// # Arguments:
-///
-/// * `metadata_spec`: is one of the const `MetadataSpec` instances from the ObjectModel trait, for the target metadata. Whether the metadata is in-header or on-side is a VM-specific choice.
-/// * `object`: is a reference to the target object.
-/// * `old_val`: is the expected current value of the metadata.
-/// * `new_val`: is the new metadata value to be stored if the compare-and-exchange operation is successful.
-/// * `mask`: is an optional mask value for the metadata. This value is used in cases like the forwarding pointer metadata, where some of the bits are reused by other metadata such as the forwarding bits.
-/// * `success_order`: is the atomic ordering used if the operation is successful.
-/// * `failure_order`: is the atomic ordering used if the operation fails.
-///
-/// # Returns `true` if the operation is successful, and `false` otherwise.
-///
-#[inline(always)]
-pub fn compare_exchange_metadata<VM: VMBinding>(
-    metadata_spec: &MetadataSpec,
-    object: ObjectReference,
-    old_val: usize,
-    new_val: usize,
-    mask: Option<usize>,
-    success_order: Ordering,
-    failure_order: Ordering,
-) -> bool {
-    match metadata_spec {
-        MetadataSpec::OnSide(metadata_spec) => side_metadata::compare_exchange_atomic(
-            metadata_spec,
-            object.to_address(),
-            old_val,
-            new_val,
-            success_order,
-            failure_order,
-        ),
-        MetadataSpec::InHeader(metadata_spec) => VM::VMObjectModel::compare_exchange_metadata(
-            metadata_spec,
-            object,
-            old_val,
-            new_val,
-            mask,
-            success_order,
-            failure_order,
-        ),
+    /// A function to atomically compare-and-exchange the specified metadata's content.
+    ///
+    /// # Arguments:
+    ///
+    /// * `metadata_spec`: is one of the const `MetadataSpec` instances from the ObjectModel trait, for the target metadata. Whether the metadata is in-header or on-side is a VM-specific choice.
+    /// * `object`: is a reference to the target object.
+    /// * `old_val`: is the expected current value of the metadata.
+    /// * `new_val`: is the new metadata value to be stored if the compare-and-exchange operation is successful.
+    /// * `mask`: is an optional mask value for the metadata. This value is used in cases like the forwarding pointer metadata, where some of the bits are reused by other metadata such as the forwarding bits.
+    /// * `success_order`: is the atomic ordering used if the operation is successful.
+    /// * `failure_order`: is the atomic ordering used if the operation fails.
+    ///
+    /// # Returns `true` if the operation is successful, and `false` otherwise.
```

Comment:
Cmpxchg should return the old value and whether it is successful or not.  The `AtomicU8::compare_exchange` method returns `Result<u8, u8>`. We should do it similarly here.  Some algorithm can benefit from it and eliminate a subsequent load.

Reply:
I will change this to `Result<T, T>`. For its use cases, I will leave it as `compare_exchange().is_ok()`.